todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java

```diff
@@ -0,0 +1,214 @@
+import com.example.android.architecture.blueprints.todoapp.util.ActivityUtils;
+
+/**
+ * Class that creates fragments (MVP views) and makes the necessary connections between them.
+ */
+public class TasksMvpController {
+
+    private final FragmentActivity mFragmentActivity;
+
+    private TasksTabletPresenter mTasksTabletPresenter;
+
+    private TasksPresenter mTasksPresenter;
+
+    private TaskDetailPresenter mTaskDetailPresenter;
+
+    /**
+     * Creates a controller for a task view for phones or tablets.
+     * @param fragmentActivity the context activity
+     * @return a TasksMvpController
+     */
+    public static TasksMvpController createTasksView(@NonNull FragmentActivity fragmentActivity) {
+        checkNotNull(fragmentActivity);
+        TasksMvpController tasksMvpController = new TasksMvpController(fragmentActivity);
+        tasksMvpController.initTasksView();
+        return tasksMvpController;
+    }
+
+    /**
+     * Creates a controller for a task detail view for phones or tablets.
+     * @param fragmentActivity the context activity
+     * @return a TasksMvpController
+     */
+    public static TasksMvpController createTaskDetailView(
+            @NonNull FragmentActivity fragmentActivity, String taskId) {
+        checkNotNull(fragmentActivity);
+        TasksMvpController tasksMvpController = new TasksMvpController(fragmentActivity);
+        tasksMvpController.initTaskDetailView(taskId);
+        return tasksMvpController;
+    }
+
+    // Force factory method, prevent direct instantiation:
+    private TasksMvpController(@NonNull FragmentActivity fragmentActivity) {
+        mFragmentActivity = fragmentActivity;
+    }
+
+    private void initTasksView() {
+        if (isTablet()) {
+            createTabletElements();
+        } else {
+            TasksFragment tasksFragment = findOrCreateTasksFragment();
+            if (tasksFragment.getPresenter() != null) {
+                mTasksPresenter = (TasksPresenter) tasksFragment.getPresenter();
+            } else {
+                mTasksPresenter = createListPresenter(tasksFragment);
+            }
+            tasksFragment.setPresenter(mTasksPresenter);
+        }
+    }
+
+    private void initTaskDetailView(String taskId) {
+        TaskDetailFragment taskDetailFragment = findOrCreateTaskDetailFragment(taskId);
+        mTaskDetailPresenter = createTaskDetailPresenter(taskId, taskDetailFragment);
+        taskDetailFragment.setPresenter(mTaskDetailPresenter);
+    }
+
+    public TasksPresenter getTasksPresenter() {
+        return mTasksPresenter;
+    }
+
+    public TaskDetailPresenter getTaskDetailPresenter() {
+        return mTaskDetailPresenter;
+    }
+
+    /**
+     * To be called in tablet mode when a new detail view is requested.
+     * @param taskId the ID of the task
+     * @return the fragment to be added to the UI
+     */
+    public TaskDetailFragment createDetailViewForTablet(String taskId) {
```

Comment:
this is called from a `Navigator` class, I don't think it's its responsibility to create `Fragments`. The reason why it's called is because instead of refreshing the `Task` that is being shown we simple recreate the `Fragment`. What do you think about making the `DetailFragment` able to refresh the task it's displaying?